PWGHF/Tasks/taskMcEfficiency.cxx

```diff
@@ -0,0 +1,290 @@
+  ConfigurableAxis axisPdg{"axisPdg", {VARIABLE_WIDTH, -421.5, 0, 421.5}, "PDG code axis"};
+  ConfigurableAxis axisCPA{"axisCPA", {100, 0.9, 1.0001}, "Cosine of pointing angle axis"};
+  Configurable<std::vector<int>> pdgCodes{"pdgCodes", {pdg::kD0Bar, pdg::kD0}, "PDG codes to process"};
+
+  Configurable<float> mcAcceptancePt{"mcAcceptancePt", 0.1, "MC Acceptance: lower pt limit"};
+  Configurable<float> mcAcceptanceEta{"mcAcceptanceEta", 0.8, "MC Acceptance: upper eta limit"};
+
+  using TracksWithSelection = soa::Join<aod::Tracks, aod::TrackSelection>;
+  using TracksWithSelectionMC = soa::Join<aod::Tracks, aod::TracksExtra, aod::McTrackLabels, aod::TrackSelection>;
+
+  void init(o2::framework::InitContext&)
+  {
+    registry.add("hCandidates", "Candidate count at different steps", {HistType::kStepTHnF, {axisPt, axisMass, axisPdg, axisCPA, {2, -0.5, 1.5, "collision matched"}}, kHFNSteps});
+    auto hCandidates = registry.get<StepTHn>(HIST("hCandidates"));
+    hCandidates->GetAxis(0)->SetTitle("p_{T} (GeV/c)");
+    hCandidates->GetAxis(1)->SetTitle("m_{inv}");
+    hCandidates->GetAxis(2)->SetTitle("PDG code");
+    hCandidates->GetAxis(3)->SetTitle("CPA");
+
+    registry.add("hDuplicateCount", "Duplicate count;frequency;count", {HistType::kTH1F, {{10, 0.5, 10.5}}});
+
+    registry.add("hTrackablePtEta", "Prongs kinematics at different steps", {HistType::kStepTHnF, {{20, 0, 10}, {40, -2, 2}}, kNTrackableSteps});
+    auto hTrackablePtEta = registry.get<StepTHn>(HIST("hCandidates"));
+    hTrackablePtEta->GetAxis(0)->SetTitle("p_{T} (GeV/c)");
+    hTrackablePtEta->GetAxis(1)->SetTitle("#eta");
+  }
+
+  template <typename T>
+  inline bool checkTrack(T track)
+  {
+    // TODO configurable?
+    return track.isGlobalTrackWoDCA();
+  }
+
+  template <bool mc, typename T1, typename T2, typename T3>
+  void candidateLoop(T1& candidates, T2& tracks, T3& mcParticles)
+  {
+    using TracksType = std::decay_t<decltype(tracks)>;
+
+    auto hCandidates = registry.get<StepTHn>(HIST("hCandidates"));
+
+    for (auto pdgCode : (std::vector<int>)pdgCodes) {
+      auto decayType = -1;
+      std::array<int, 2> pdgDaughters;
+
+      if (pdgCode == pdg::kD0) {
+        decayType = 1 << DecayType::D0ToPiK;
+        pdgDaughters[0] = +kPiPlus;
+        pdgDaughters[1] = -kKPlus;
+      } else if (pdgCode == pdg::kD0Bar) {
+        decayType = 1 << DecayType::D0ToPiK;
+        pdgDaughters[0] = -kPiPlus;
+        pdgDaughters[1] = +kKPlus;
+      } else {
+        LOGP(fatal, "Not implemented for PDG {}", pdgCode);
+      }
+
+      std::map<int64_t, int> duplicates;
+
+      for (auto& candidate : selectedD0Candidates) {
+        if (!(candidate.hfflag() & decayType)) {
+          continue;
+        }
+
+        bool collisionMatched = false;
+        if constexpr (mc) {
+          auto trackPos = candidate.prong0_as<TracksType>();
+          auto trackNeg = candidate.prong1_as<TracksType>();
```

Comment:
These can be taken out of the `mc` scope and reused later in the `checkTrack` calls.